src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs

```diff
@@ -213,5 +235,36 @@ private bool TryReadArgumentsAsNamedValues(out ICollection<FunctionParameterToke
 
             return false;
         }
+
+        /// <summary>
+        /// Set the parent of the next argument to the current argument.
+        /// For example, in the following query:
+        ///     cast(Location, NS.HomeAddress) where Location is the parentExpression and NS.HomeAddress is the parameterToken.
+        ///     isof(Location, NS.HomeAddress) where Location is the parentExpression and NS.HomeAddress is the parameterToken.
+        /// </summary>
+        /// <param name="parentExpression">The previous parameter token.</param>
+        /// <param name="parameterToken">The current parameter token.</param>
+        /// <returns>The updated parameter token.</returns>
+        private static QueryToken SetParentForCurrentParameterToken(QueryToken parentExpression, QueryToken parameterToken)
+        {
+            if (parameterToken is not DottedIdentifierToken dottedIdentifierToken || dottedIdentifierToken?.NextToken is not null)
+            {
+                return parameterToken;
+            }
+
+            // Check if the identifier is a primitive type
+            IEdmSchemaElement schemaElement = EdmCoreModel.Instance.SchemaElements

```

Comment:
Is there where it's been suggested to do a cache look-up? How many elements do we have in the core model schema elements?